### PR TITLE
xds: wrr with random

### PR DESCRIPTION
**Pull request by menghanl**
for drops and inter-locality picks

### CODE REVIEW

File: balancer/internal/wrr/random.go [hunk 1]
```diff
@@ -0,0 +1,58 @@
...
+// NewRandom creates a new WRR with random.
+func NewRandom() WRR {
+	return &randomWRR{}
+}
+
+func (rw *randomWRR) Next() (item interface{}) {
+	if rw.sumOfWeights == 0 {
+		return nil
+	}
+	randomWeight := grpcrand.Int63n(rw.sumOfWeights)
```
Comment (lyuxuan) [inline balancer/internal/wrr/random.go]: Make a comment saying that grpcrand.Int63n generates an integer [0, rw.sumOfweights).

File: balancer/internal/wrr/random.go [hunk 2]
```diff
@@ -0,0 +1,58 @@
...
+	for _, item := range rw.items {
+		randomWeight = randomWeight - item.Weight
+		if randomWeight < 0 {
+			return item.Item
+		}
+	}
+
+	return rw.items[len(rw.items)-1].Item
```
Comment (lyuxuan) [inline balancer/internal/wrr/random.go]: I think this is not necessary. The for loop will cover all the cases.

File: balancer/internal/wrr/random.go [hunk 3]
```diff
@@ -0,0 +1,58 @@
...
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package wrr
+
+import "google.golang.org/grpc/internal/grpcrand"
+
+// randWeighted is a wrapped weighted item that is used to implement weighted random algorithm.
+type randWeighted struct {
```
Comment (lyuxuan) [inline balancer/internal/wrr/random.go]: can we remove the "rand" prefix, and name it e.g. `weightedItem`? The struct itself is unrelated to randomness, it's just you currently only used it for the `randomWRR`.

File: balancer/xds/edsbalancer/util_test.go [hunk 4]
```diff
@@ -20,8 +20,49 @@ package edsbalancer
...
 	"testing"
+
+	"google.golang.org/grpc/balancer/internal/wrr"
 )
 
+// testWRR is a deterministic WRR implementation.
+//
+// With {a: 2, b: 3}, the Next() results will be {a, a, b, b, b}.
+type testWRR struct {
+	itemsWithWeight []struct {
```
Comment (lyuxuan) [inline balancer/xds/edsbalancer/util_test.go]: You can reuse the `randWeighted` struct once you change the name.

File: balancer/xds/edsbalancer/util_test.go [hunk 5]
```diff
@@ -20,8 +20,49 @@ package edsbalancer
 
 import (
 	"testing"
+
+	"google.golang.org/grpc/balancer/internal/wrr"
 )
 
+// testWRR is a deterministic WRR implementation.
```
Comment (lyuxuan) [inline balancer/xds/edsbalancer/util_test.go]: Can you explain in the comment why we need the deterministic WRR implementation for testing?

File: balancer/internal/wrr/random.go (hunk 3 above)
Comment (menghanl) [reply balancer/internal/wrr/random.go]: Done

File: balancer/internal/wrr/random.go (hunk 1 above)
Comment (menghanl) [reply balancer/internal/wrr/random.go]: Done

File: balancer/internal/wrr/random.go (hunk 2 above)
Comment (menghanl) [reply balancer/internal/wrr/random.go]: Compiler will think this is missing a return.

File: balancer/internal/wrr/wrr.go [hunk 6]
```diff
@@ -0,0 +1,26 @@
...
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package wrr
+
+// WRR blah
```
[the remark this answers is not in the document]
Comment (menghanl) [reply balancer/internal/wrr/wrr.go]: Done

File: balancer/xds/edsbalancer/util_test.go (hunk 5 above)
Comment (menghanl) [reply balancer/xds/edsbalancer/util_test.go]: Done

File: balancer/xds/edsbalancer/util_test.go (hunk 4 above)
Comment (menghanl) [reply balancer/xds/edsbalancer/util_test.go]: They are in different packages.

File: balancer/xds/edsbalancer/util_test.go [hunk 7]
```diff
@@ -20,8 +20,50 @@ package edsbalancer
 
 import (
 	"testing"
+
+	"google.golang.org/grpc/balancer/internal/wrr"
 )
 
+// testWRR is a deterministic WRR implementation. With this, the balancer
```
[the remark this answers is not in the document]
Comment (menghanl) [reply balancer/xds/edsbalancer/util_test.go]: Done

File: balancer/internal/wrr/wrr.go [hunk 8]
```diff
@@ -0,0 +1,26 @@
...
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package wrr
+
+// WRR defines an interface that implements weighted round robin.
+type WRR interface {
```
Comment (lyuxuan) [inline balancer/internal/wrr/wrr.go]: Does WRR need to be thread safe? Or if there's a guarantee that Add or Next won't be called concurrently. It's better to document here clearly.

Comment (menghanl) [reply balancer/internal/wrr/wrr.go]: Next() needs to be thread safe.

Add() I'm not sure. I'm leaving it undocumented. Will add more requirements when adding new implementations.

**Pull request merged**